### PR TITLE
Integrated Google Cloud Storage

### PR DESCRIPTION
**Pull request by Korusuke**
Co-authored-by: korusuke <karan.sheth@somaiya.edu>

## Description

Adds Google Cloud Storage (GCS) support for storing BentoML service, this is an alternative to S3/MiniIO.

This internally changes `s3_presigned_url` to `cloud_presigned_url`, this will allow for better readability as more cloud storage can be supported in future. No existing S3 functionality is affected/changed due to this PR.

## Motivation and Context

#661 

## How Has This Been Tested?

Manually tested
End-to-end tests are added, but they were not tested cause of storage issues😅

## Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] New feature and improvements (non-breaking change which adds/improves functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Code Refactoring (internal change which is not user facing)
- [ ] Documentation
- [x] Test, CI, or build

## Component(s) if applicable
- [x] BentoService (service definition, dependency management, API input/output adapters)
- [ ] Model Artifact (model serialization, multi-framework support)
- [ ] Model Server (mico-batching, dockerisation, logging, OpenAPI, instruments)
- [x] YataiService gRPC server (model registry, cloud deployment automation)
- [ ] YataiService web server (nodejs HTTP server and web UI)
- [ ] Internal (BentoML's own configuration, logging, utility, exception handling)
- [ ] BentoML CLI

## Checklist:

- [x] My code follows the bentoml code style, both `./dev/format.sh` and
  `./dev/lint.sh` script have passed
  ([instructions](https://github.com/bentoml/BentoML/blob/master/DEVELOPMENT.md#style-check-and-auto-formatting-your-code)).
- [ ] My change reduces project test coverage and requires unit tests to be added
- [ ] I have added unit tests covering my code change
- [x] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly

### CODE REVIEW

**Comment by yubozhao** [comment]
Hi, @Korusuke I talked with @PrabhanshuAttri Let's add unit tests and e2e test for Google cloud storage.  This is a very critical piece of Yatai service and a lot of people are depending on this. Want to be more deliberate with this.

File: bentoml/cli/bento_service.py [hunk 1]
```diff
@@ -120,7 +121,7 @@ def resolve_bundle_path(bento, pip_installed_bundle_path):
         ), "pip installed BentoService commands should not have Bento argument"
         return pip_installed_bundle_path
 
-    if os.path.isdir(bento) or is_s3_url(bento):
+    if os.path.isdir(bento) or is_s3_url(bento) or is_gcs_url(bento):
         # saved_bundle already support loading local and s3 path
```
Comment (yubozhao) [inline bentoml/cli/bento_service.py:125]: Update the comment to reflect this change

File: bentoml/saved_bundle/loader.py [hunk 2]
```diff
@@ -55,10 +58,22 @@ def _resolve_remote_bundle_path(bundle_path):
...
+        fileobj.seek(0, 0)
     elif _is_http_url(bundle_path):
         import requests
 
         response = requests.get(bundle_path)
+        if response.status_code != 200:
+            raise BentoMLException(
+                f"Error retrieving BentoService bundle. "
+                f"{response.status_code}: {response.text}"
+            )
```
Comment (yubozhao) [inline bentoml/saved_bundle/loader.py:76]: 👍

File: bentoml/utils/gcs.py [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+
+import logging
+from urllib.parse import urlparse
+
+logger = logging.getLogger(__name__)
+
+
+def is_gcs_url(url):
+    """
+    Check if url is an gs url
```
Comment (yubozhao) [inline bentoml/utils/gcs.py:23]: Check if the URL is a GCS URL.

File: bentoml/utils/gcs.py [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+    """
+    try:
+        return urlparse(url).scheme in ["gs"]
```
Comment (yubozhao) [inline bentoml/utils/gcs.py:26]: Is `gs://` the standard way for Google cloud URI?  Could we document this, if you have a reference?

File: bentoml/yatai/client/bento_repository_api.py [hunk 5]
```diff
@@ -146,6 +146,38 @@ def _upload_bento_service(self, bento_service, saved_bento_path):
...
+
+            logger.info(
+                "Successfully saved BentoService bundle '%s:%s' to GCS: %s",
+                bento_service_metadata.name,
+                bento_service_metadata.version,
+                response.uri.uri,
+            )
+
+            return response.uri.uri
+
```
Comment (yubozhao) [inline bentoml/yatai/client/bento_repository_api.py:180]: This seems like using the same code path as the S3 one.  Let's refactor this.

File: bentoml/yatai/yatai_service_impl.py [hunk 6]
```diff
@@ -398,7 +396,11 @@ def GetBento(self, request, context=None):
...
             if bento_pb:
                 if bento_pb.uri.type == BentoUri.S3:
-                    bento_pb.uri.s3_presigned_url = self.repo.get(
+                    bento_pb.uri.cloud_presigned_url = self.repo.get(
+                        bento_pb.name, bento_pb.version
+                    )
+                if bento_pb.uri.type == BentoUri.GCS:
+                    bento_pb.uri.cloud_presigned_url = self.repo.get(
                         bento_pb.name, bento_pb.version
                     )
```
Comment (yubozhao) [inline bentoml/yatai/yatai_service_impl.py:405]: Let's refactor these two.

Comment (PrabhanshuAttri) [reply bentoml/yatai/yatai_service_impl.py:405]: @yubozhao As far as I can understand, we should merge these 2 if conditions. Please confirm.

File: bentoml/yatai/client/bento_repository_api.py (hunk 5 above)
Comment (PrabhanshuAttri) [reply bentoml/yatai/client/bento_repository_api.py:180]: @yubozhao Would you like us to merge this `if` condition with S3 `if` condition?

File: bentoml/utils/gcs.py (hunk 4 above)
Comment (PrabhanshuAttri) [reply bentoml/utils/gcs.py:26]: @yubozhao Should we put the documentation link as a comment?

Here is the documentation for the same https://cloud.google.com/storage/docs/gsutil

File: bentoml/yatai/yatai_service_impl.py (hunk 6 above)
Comment (yubozhao) [reply bentoml/yatai/yatai_service_impl.py:405]: Yes we should refactor these two if statement.

File: bentoml/yatai/client/bento_repository_api.py (hunk 5 above)
Comment (yubozhao) [reply bentoml/yatai/client/bento_repository_api.py:180]: Yes.  The only difference between those two statements is raising exceptions. We should be refactoring them

File: bentoml/yatai/yatai_service_impl.py (hunk 6 above)
Comment (PrabhanshuAttri) [reply bentoml/yatai/yatai_service_impl.py:405]: Got it.

File: bentoml/utils/gcs.py (hunk 4 above)
Comment (yubozhao) [reply bentoml/utils/gcs.py:26]: Ok great. Would you mind include that link in the form of comments for this funciton?

**Review by yubozhao — DISMISSED** [review]

File: bentoml/saved_bundle/loader.py [hunk 7]
```diff
@@ -55,10 +58,22 @@ def _resolve_remote_bundle_path(bundle_path):
         fileobj = io.BytesIO()
         s3.download_fileobj(bucket_name, object_name, fileobj)
         fileobj.seek(0, 0)
+    elif is_gcs_url(bundle_path):
+        from google.cloud import storage
```
Comment (parano) [inline bentoml/saved_bundle/loader.py:62]: handle import error

File: bentoml/yatai/repository/gcs_repository.py [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+
+
+class GCSRepository(BaseRepository):
+    def __init__(self, base_url):
+        try:
+            from google.cloud import storage
+        except ImportError:
+            raise YataiRepositoryException(
+                '"google-cloud-storage" package is required for Google Cloud '
+                'Storage Repository.'
```
Comment (parano) [inline bentoml/yatai/repository/gcs_repository.py:33]: add `pip install XXXX` to the error message

Comment (parano) [reply bentoml/yatai/repository/gcs_repository.py:33]: it will be helpful when a user is using local yatai

File: protos/repository.proto [hunk 9]
```diff
@@ -19,7 +19,7 @@ message BentoUri {
 
   StorageType type = 1;
   string uri = 2;
-  string s3_presigned_url = 3;
```
Comment (parano) [inline protos/repository.proto:22]: we should avoid this in the future, especially given more users are using remote YataiService and there will be backward compatibility issues

File: setup.py [hunk 10]
```diff
@@ -95,7 +93,11 @@
 
 dev_all = install_requires + dev_requires + docs_requires
 
-yatai_service = install_requires + aws_sam_cli + postgres + azure_cli
+yatai_service = (
+    ["google-cloud-storage", "azure-cli", "psycopg2", "psycopg2-binary"]
+    + install_requires
+    + aws_sam_cli
```
Comment (parano) [inline setup.py:99]: @yubozhao  is aws_sam_cli still used in tests?  can we get rid of it in unit tests?

**Pull request merged**